test_regress/t/t_vpi_package.cpp

```diff
@@ -0,0 +1,216 @@
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+//*************************************************************************
+
+#ifdef IS_VPI
+
+#include "sv_vpi_user.h"
+
+#include <cstdlib>
+
+#else
+
+#include "verilated.h"
+#include "verilated_vcd_c.h"
+#include "verilated_vpi.h"
+
+#include "Vt_vpi_package.h"
+#include "Vt_vpi_package__Dpi.h"
+#include "svdpi.h"
+
+#endif
+
+#include <cstdio>
+#include <cstring>
+#include <iostream>
+
+// These require the above. Comment prevents clang-format moving them
+#include "TestSimulator.h"
+#include "TestVpi.h"
+
+// __FILE__ is too long
+#define FILENM "t_vpi_package.cpp"
+
+#define DEBUG \
+    if (0) printf
+
+#define CHECK_RESULT_NZ(got) \
+    if (!(got)) { \
+        printf("%%Error: %s:%d: GOT = NULL  EXP = !NULL\n", FILENM, __LINE__); \
+        return __LINE__; \
+    }
+
+#define CHECK_RESULT_Z(got) \
+    if (got) { \
+        printf("%%Error: %s:%d: GOT = !NULL  EXP = NULL\n", FILENM, __LINE__); \
+        return __LINE__; \
+    }
+
+#define CHECK_RESULT(got, exp) \
+    if ((got) != (exp)) { \
+        std::cout << std::dec << "%Error: " << FILENM << ":" << __LINE__ << ": GOT = " << (got) \
+                  << "   EXP = " << (exp) << std::endl; \
+        return __LINE__; \
+    }
+
+#define CHECK_RESULT_CSTR(got, exp) \
+    if (std::strcmp((got), (exp))) { \
+        printf("%%Error: %s:%d: GOT = '%s'   EXP = '%s'\n", FILENM, __LINE__, \
+               (got) ? (got) : "<null>", (exp) ? (exp) : "<null>"); \
+        return __LINE__; \
+    }
+
+extern "C" {
+int mon_check() {
+#ifdef TEST_VERBOSE
+    printf("-mon_check()\n");
+#endif
+
+    vpiHandle it = vpi_iterate(vpiModule, NULL);
+
+    bool found_t = false;
+    while (true) {
+        vpiHandle handle = vpi_scan(it);
+        if (handle == NULL) break;
+        if (strcmp("t", vpi_get_str(vpiName, handle))) {
+            return __LINE__;
+        } else {
+            if (found_t) return __LINE__;
+            found_t = true;
+        }
+    }
+    if (!found_t) return __LINE__;
+
+    it = vpi_iterate(vpiInstance, NULL);
+
+    found_t = false;
+    bool found_somepackage = false;
+    bool found_dollar_unit = false;
+    while (true) {
+        vpiHandle handle = vpi_scan(it);
+        if (handle == NULL) break;
+        const char* name = vpi_get_str(vpiName, handle);
+        const char* fullname = vpi_get_str(vpiFullName, handle);
+        if (!strcmp("t", name)) {
+            if (strcmp("t", fullname)) return __LINE__;
+            if (found_t) return __LINE__;
+            found_t = true;
+        } else if (!strcmp("somepackage", name)) {
+            if (strcmp("somepackage::", fullname)) return __LINE__;
+            if (found_somepackage) return __LINE__;
+            found_somepackage = true;
+            // NOCOMMIT -- fix names
+        } else if (!strcmp("\\$unit ", name)) {
+            if (strcmp("\\$unit ::", fullname)) return __LINE__;
```

Comment:
The scope name for `$unit` is `\$unit ` (with a trailing space).  I'm going to try to fix this, but is there some reason it is this way?

Reply:
The $ needs escaping because it's otherwise illegal as a Verilog symbol.  Once escaped, escaped identifiers need to end with a space.   It might be that other simulators ignore that and just use the $ directly, but I suspect you will then run into issues that the Verilator name quoting rules don't like it being inconsistent.  So anyhow, be prepared this might be a much more complicated issue than appears.

Reply:
What about the trailing space?  Is that intentional / meaningful as well?

Also, how do you feel about adding a special case to `name()` and `fullname()` in `VerilatedVpioPackage`?  That feels like perhaps the easiest way to tackle this.

And yes, other simulators do **all** sorts of other things.